In this particular notebook, we're gonna learn about the control over missing values in our datasets.<br>
We are going to use the dataset from [here](https://www.kaggle.com/dansbecker/melbourne-housing-snapshot) 

In [1]:
# so let's load our dataset
import pandas as pd 
from sklearn.model_selection import train_test_split

data = pd.read_csv('../input/melbourne-housing-snapshot/melb_data.csv')

# selected target to drop this column
y = data.Price

Have a check the columns of our dataset

In [2]:
print(data.columns)
print("column size:", data.columns.size)

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')
column size: 21


Let's drop the *Price* column which was our target.

In [3]:
melb_predictors = data.drop(['Price'], axis = 1)

See the new dataset has everything except *Price* column. We can have a look to the size of the columns now, it's 20!

In [4]:
print(melb_predictors.columns)
print("column size:", melb_predictors.columns.size)

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Method', 'SellerG', 'Date',
       'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car', 'Landsize',
       'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude', 'Longtitude',
       'Regionname', 'Propertycount'],
      dtype='object')
column size: 20


Now, let's make our dataset more interesting by ignoring strings<br>
as string have less often efects on our model (for such value based predictions particularly)<br>
**Note:** String data type here is *object*





In [5]:
# let's print out the data types of each column
for col in melb_predictors.columns:
    print(col, data[col].dtype)

Suburb object
Address object
Rooms int64
Type object
Method object
SellerG object
Date object
Distance float64
Postcode float64
Bedroom2 float64
Bathroom float64
Car float64
Landsize float64
BuildingArea float64
YearBuilt float64
CouncilArea object
Lattitude float64
Longtitude float64
Regionname object
Propertycount float64


In [6]:
# let's ignore the columns having strings (object dtype)
X = melb_predictors.select_dtypes(exclude = ['object'])

In [7]:
# let's see the coulmns
print(X.columns)
print("column size:", X.columns.size)

Index(['Rooms', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'Lattitude', 'Longtitude',
       'Propertycount'],
      dtype='object')
column size: 12


Well, now the dataset has only 12 columns. Have a look at the dataframe...

In [8]:
X

Rooms  Distance  Postcode  Bedroom2  Bathroom  Car  Landsize  \
0          2       2.5    3067.0       2.0       1.0  1.0     202.0   
1          2       2.5    3067.0       2.0       1.0  0.0     156.0   
2          3       2.5    3067.0       3.0       2.0  0.0     134.0   
3          3       2.5    3067.0       3.0       2.0  1.0      94.0   
4          4       2.5    3067.0       3.0       1.0  2.0     120.0   
...      ...       ...       ...       ...       ...  ...       ...   
13575      4      16.7    3150.0       4.0       2.0  2.0     652.0   
13576      3       6.8    3016.0       3.0       2.0  2.0     333.0   
13577      3       6.8    3016.0       3.0       2.0  4.0     436.0   
13578      4       6.8    3016.0       4.0       1.0  5.0     866.0   
13579      4       6.3    3013.0       4.0       1.0  1.0     362.0   

       BuildingArea  YearBuilt  Lattitude  Longtitude  Propertycount  
0               NaN        NaN  -37.79960   144.99840         4019.0  
1              79.0     1900.0  -37.80790   144.99340         4019.0  
2             150.0     1900.0  -37.80930   144.99440         4019.0  
3               NaN        NaN  -37.79690   144.99690         4019.0  
4             142.0     2014.0  -37.80720   144.99410         4019.0  
...             ...        ...        ...         ...            ...  
13575           NaN     1981.0  -37.90562   145.16761         7392.0  
13576         133.0     1995.0  -37.85927   144.87904         6380.0  
13577           NaN     1997.0  -37.85274   144.88738         6380.0  
13578         157.0     1920.0  -37.85908   144.89299         6380.0  
13579         112.0     1920.0  -37.81188   144.88449         6543.0  

[13580 rows x 12 columns]

But, here are so many missing values in our dataset and which is a problem! :(<br>
Forget this, and more forward. Let's split the current dataset into **train** and **validation** set.

In [9]:
X_train, X_valid, y_train, y_valid = train_test_split(X, y, train_size = .8, test_size = .2, random_state = 0)

As most often we need to calculate our mean_absolute_error function, we're gonna define a function for it here!<br>
**Note:** *n_estimators* is the number of trees you want to build before taking the maximum voting or averages of predictions of random forest.<br>
Or, you may think it as the number of trees consisting the random forest.

In [10]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

# function returen absolute mean error of two columns
def score_dataset(X_train, X_valid, y_train, y_valid):
    model = RandomForestRegressor(n_estimators = 10, random_state = 0)
    model.fit(X_train, y_train)
    preds = model.predict(X_valid)
    return mean_absolute_error(y_valid, preds)
    
    

Now we're gonna deal with the missing values<br>
# **Approach 1:** Drop Columns with Missing values

In [11]:
# just have a look to the datasetn to see NaN or missing values
X_train

Rooms  Distance  Postcode  Bedroom2  Bathroom  Car  Landsize  \
12167      1       5.0    3182.0       1.0       1.0  1.0       0.0   
6524       2       8.0    3016.0       2.0       2.0  1.0     193.0   
8413       3      12.6    3020.0       3.0       1.0  1.0     555.0   
2919       3      13.0    3046.0       3.0       1.0  1.0     265.0   
6043       3      13.3    3020.0       3.0       1.0  2.0     673.0   
...      ...       ...       ...       ...       ...  ...       ...   
13123      3       5.2    3056.0       3.0       1.0  2.0     212.0   
3264       3      10.5    3081.0       3.0       1.0  1.0     748.0   
9845       4       6.7    3058.0       4.0       2.0  2.0     441.0   
10799      3      12.0    3073.0       3.0       1.0  1.0     606.0   
2732       4       6.4    3011.0       4.0       2.0  1.0     319.0   

       BuildingArea  YearBuilt  Lattitude  Longtitude  Propertycount  
12167           NaN     1940.0  -37.85984   144.98670        13240.0  
6524            NaN        NaN  -37.85800   144.90050         6380.0  
8413            NaN        NaN  -37.79880   144.82200         3755.0  
2919            NaN     1995.0  -37.70830   144.91580         8870.0  
6043          673.0     1970.0  -37.76230   144.82720         4217.0  
...             ...        ...        ...         ...            ...  
13123           NaN        NaN  -37.77695   144.95785        11918.0  
3264          101.0     1950.0  -37.74160   145.04810         2947.0  
9845          255.0     2002.0  -37.73572   144.97256        11204.0  
10799           NaN        NaN  -37.72057   145.02615        21650.0  
2732          130.0     1915.0  -37.79430   144.88750         7570.0  

[10864 rows x 12 columns]

In [12]:
# Fist we'll see how it works for a single column 
# actually, we're going to see if any missing value(NULL or NaN) present or not
# this line of code below represents missing values as False
print(X_train['BuildingArea'].isnull())
# we can also print the numner of missing values
print("Total missing values", sum(X_train['BuildingArea'].isnull()))

12167     True
6524      True
8413      True
2919      True
6043     False
         ...  
13123     True
3264     False
9845     False
10799     True
2732     False
Name: BuildingArea, Length: 10864, dtype: bool
Total missing values 5156


See! Mssing values showed as *True* and others have *False*<br>
and there are 5156 missing values in this column. Now, just use *.any()* which shows **True** if *missing values > 0*

In [13]:
X_train["BuildingArea"].isnull().any()

True

**Yes**! Let's do it on our dataset, right!

In [14]:
# you're gonna get the names of the columns containing any single missing value
cols_with_missing = [col for col in X_train.columns if X_train[col].isnull().any()]
cols_with_missing

['Car', 'BuildingArea', 'YearBuilt']

See,  these three columns are containing one or more missing values.<br>
So, just drop those columns and calculate *mean_absolute_error* again.

In [15]:
reduced_X_train = X_train.drop(cols_with_missing, axis = 1)
reduced_X_valid = X_valid.drop(cols_with_missing, axis = 1)
# let's print how many columns are ramaning now!
reduced_X_train.columns.size

9

In [16]:
# so let's ca;culate the mean error
print("MAE from Approach 1: Drop Columns with Missing values")
print(score_dataset(reduced_X_train, reduced_X_valid, y_train, y_valid))

MAE from Approach 1: Drop Columns with Missing values
183550.22137772635


Now,  we're going to see the second approach.
# **Approach 2**: Imputation

To learn about how imputation actually works, just go through this link https://scikit-learn.org/stable/modules/impute.html<br>
However, we're going to practice it before implement as it's kinda new to us, right!

In [17]:
# let's import it from sklearn libranry
from sklearn.impute import SimpleImputer
import numpy as np
# declare a numpy array
train_array = np.array([[1, 2], [np.nan, 3], [2, 3]])
train_array

array([[ 1.,  2.],
       [nan,  3.],
       [ 2.,  3.]])

See, in our numpy array, we have the 'nan' as missing value.<br>
Let's define an imputer on mssing data with the strategy *mean*

In [18]:
imp = SimpleImputer(missing_values = np.nan, strategy = 'mean')
# and fit the imputer on train_array we have just declared
imp.fit(train_array)

SimpleImputer()

So, we see the imputer has been created as **SimpleImputer()**<br>
As the imputer creation is done, we can  use it for our further use on others.

In [19]:
# so now we're gonna transfornm our train_array with the inmputer ans replace 'nan' with mean of the entire column
imp.transform(train_array)

array([[1. , 2. ],
       [1.5, 3. ],
       [2. , 3. ]])

As we already have the **SimpleImputer**() let's use it on another data

In [20]:
test_array = np.array([[np.nan, 2], [6, np.nan], [7, 6]])
test_array

array([[nan,  2.],
       [ 6., nan],
       [ 7.,  6.]])

In [21]:
imp.transform(test_array)

array([[1.5       , 2.        ],
       [6.        , 2.66666667],
       [7.        , 6.        ]])

Simliarly, we can use this with *median* strategy

In [22]:
imp = SimpleImputer(missing_values = np.nan, strategy = 'median')
imp.fit(test_array)

SimpleImputer(strategy='median')

In [23]:
imp.transform(test_array)

array([[6.5, 2. ],
       [6. , 4. ],
       [7. , 6. ]])

Another thing, if we want both the *.fit()* and *.transform()* work together,  then we just use *fit_transform()*<br>
rather than  using them individully!

In [24]:
# we're taking the same 'test_array' which we did previously
imp = SimpleImputer(missing_values = np.nan, strategy = 'mean')
imp.fit_transform(test_array)

array([[6.5, 2. ],
       [6. , 4. ],
       [7. , 6. ]])

**Wow!** see how this works!<br>
Let's do it on our dataset on which we're working off.

In [25]:
# we're not passing anything, so the default is strategy = 'mean'
my_imputer = SimpleImputer()

In [26]:
imputed_X_train = pd.DataFrame(my_imputer.fit_transform(X_train))

But, we wont do *.fit()* on validation data as this one will be for the test purpose.

In [27]:
imputed_X_valid = pd.DataFrame(my_imputer.transform(X_valid))

In [28]:
imputed_X_train

0     1       2    3    4    5      6           7            8   \
0      1.0   5.0  3182.0  1.0  1.0  1.0    0.0  153.764119  1940.000000   
1      2.0   8.0  3016.0  2.0  2.0  1.0  193.0  153.764119  1964.839866   
2      3.0  12.6  3020.0  3.0  1.0  1.0  555.0  153.764119  1964.839866   
3      3.0  13.0  3046.0  3.0  1.0  1.0  265.0  153.764119  1995.000000   
4      3.0  13.3  3020.0  3.0  1.0  2.0  673.0  673.000000  1970.000000   
...    ...   ...     ...  ...  ...  ...    ...         ...          ...   
10859  3.0   5.2  3056.0  3.0  1.0  2.0  212.0  153.764119  1964.839866   
10860  3.0  10.5  3081.0  3.0  1.0  1.0  748.0  101.000000  1950.000000   
10861  4.0   6.7  3058.0  4.0  2.0  2.0  441.0  255.000000  2002.000000   
10862  3.0  12.0  3073.0  3.0  1.0  1.0  606.0  153.764119  1964.839866   
10863  4.0   6.4  3011.0  4.0  2.0  1.0  319.0  130.000000  1915.000000   

             9          10       11  
0     -37.85984  144.98670  13240.0  
1     -37.85800  144.90050   6380.0  
2     -37.79880  144.82200   3755.0  
3     -37.70830  144.91580   8870.0  
4     -37.76230  144.82720   4217.0  
...         ...        ...      ...  
10859 -37.77695  144.95785  11918.0  
10860 -37.74160  145.04810   2947.0  
10861 -37.73572  144.97256  11204.0  
10862 -37.72057  145.02615  21650.0  
10863 -37.79430  144.88750   7570.0  

[10864 rows x 12 columns]

see imputer has removed the column name, let's put them again

In [29]:
imputed_X_train.columns = X_train.columns
imputed_X_valid.columns = X_valid.columns

In [30]:
imputed_X_train

Rooms  Distance  Postcode  Bedroom2  Bathroom  Car  Landsize  \
0        1.0       5.0    3182.0       1.0       1.0  1.0       0.0   
1        2.0       8.0    3016.0       2.0       2.0  1.0     193.0   
2        3.0      12.6    3020.0       3.0       1.0  1.0     555.0   
3        3.0      13.0    3046.0       3.0       1.0  1.0     265.0   
4        3.0      13.3    3020.0       3.0       1.0  2.0     673.0   
...      ...       ...       ...       ...       ...  ...       ...   
10859    3.0       5.2    3056.0       3.0       1.0  2.0     212.0   
10860    3.0      10.5    3081.0       3.0       1.0  1.0     748.0   
10861    4.0       6.7    3058.0       4.0       2.0  2.0     441.0   
10862    3.0      12.0    3073.0       3.0       1.0  1.0     606.0   
10863    4.0       6.4    3011.0       4.0       2.0  1.0     319.0   

       BuildingArea    YearBuilt  Lattitude  Longtitude  Propertycount  
0        153.764119  1940.000000  -37.85984   144.98670        13240.0  
1        153.764119  1964.839866  -37.85800   144.90050         6380.0  
2        153.764119  1964.839866  -37.79880   144.82200         3755.0  
3        153.764119  1995.000000  -37.70830   144.91580         8870.0  
4        673.000000  1970.000000  -37.76230   144.82720         4217.0  
...             ...          ...        ...         ...            ...  
10859    153.764119  1964.839866  -37.77695   144.95785        11918.0  
10860    101.000000  1950.000000  -37.74160   145.04810         2947.0  
10861    255.000000  2002.000000  -37.73572   144.97256        11204.0  
10862    153.764119  1964.839866  -37.72057   145.02615        21650.0  
10863    130.000000  1915.000000  -37.79430   144.88750         7570.0  

[10864 rows x 12 columns]

see the column names are back! <br>
So, it's time to calculate the *mean_absolute_error* of current dataset

In [31]:
print("MAE from Approach 2: Imputation")
print(score_dataset(imputed_X_train, imputed_X_valid, y_train, y_valid))

MAE from Approach 2: Imputation
178166.46269899711


This is better! comparing our previous error, right! :)<br>
So, the **Approach 2** performs better than the **Approach 1** in this example.

Now, we're gonna see the third approach.<r>
# Approach 3: An Extension to Imputation

In [32]:
# making a copy of the original data
X_train_plus = X_train.copy()
X_valid_plus = X_valid.copy()

Let's make new column(additional) indicating what will be imputed

In [33]:
# .isnull will return boolian value for each entry of a column
for col in cols_with_missing:
    X_train_plus[col + '_was_missing'] = X_train_plus[col].isnull()
    X_valid_plus[col + '_was_missing'] = X_valid_plus[col].isnull()

So, again do imputation on our dataset

In [34]:
new_imputer = SimpleImputer()
imputed_X_train_plus = pd.DataFrame(new_imputer.fit_transform(X_train_plus))
imputed_X_valid_plus = pd.DataFrame(new_imputer.transform(X_valid_plus))

In [35]:
# putting the column name back again
imputed_X_train_plus.columns = X_train_plus.columns
imputed_X_valid_plus.columns = X_valid_plus.columns

Last thing is to calculate the *mean_absolute_error*. Let's do it!

In [36]:
print("MAE from Approach 3: Extension to Imputation")
print(score_dataset(imputed_X_train_plus, imputed_X_valid_plus, y_train, y_valid))

MAE from Approach 3: Extension to Imputation
178927.503183954


Ouw!! The score is not improved here than the previous score. :( <br>

**So, why did imputation perform better than dropping the columns?**<br>
The training data has 10864 rows and 12 columns, where three columns contain missing data. For each column, less than half of the entries are missing.<br>
Thus, dropping the columns removes a lot of useful information, and so it makes sense that imputation would perform better.

Finally, to wrap up. We conclude by drafting the missing values<br>
of that 3 columns containing some missing values

In [37]:
print(X_train.shape)

# number of missing values in each column of X_train
missing_val_count_by_column = (X_train.isnull().sum())
print(missing_val_count_by_column[missing_val_count_by_column > 0])

(10864, 12)
Car               49
BuildingArea    5156
YearBuilt       4307
dtype: int64


# Congratualtions!<br>
We now know how to deal with missing values in our dataset with atleast three approaches.